pkg/sentry/platform/ring0/entry_amd64.s

```diff
@@ -95,58 +89,87 @@ TEXT ·swapgs(SB),NOSPLIT,$0
 	SWAP_GS()
 	RET
 
+// jumpToKernel changes execution to the kernel address space.
+//
+// This works by changing the return value to the kernel version.
+TEXT ·jumpToKernel(SB),NOSPLIT,$0
+	MOVQ 0(SP), AX
+	ORQ ·KernelStartAddress(SB), AX // Future return value.
+	MOVQ AX, 0(SP)
+	RET
+
 // See entry_amd64.go.
 TEXT ·sysret(SB),NOSPLIT,$0-24
+	CALL ·jumpToKernel(SB)
 	// Save original state.
-	LOAD_KERNEL_ADDRESS(cpu+0(FP), BX)
-	LOAD_KERNEL_ADDRESS(regs+8(FP), AX)
+	MOVQ cpu+0(FP), BX
+	MOVQ regs+8(FP), AX
+	MOVQ userCR3+16(FP), CX
 	MOVQ SP, CPU_REGISTERS+PTRACE_RSP(BX)
 	MOVQ BP, CPU_REGISTERS+PTRACE_RBP(BX)
 	MOVQ AX, CPU_REGISTERS+PTRACE_RAX(BX)
 
+	// save SP AX userCR3 on the kernel stack
+	MOVQ CPU_ENTRY(BX), BX
+	LOAD_KERNEL_STACK(BX)
+	PUSHQ PTRACE_RSP(AX)
+	PUSHQ PTRACE_RAX(AX)
+	PUSHQ CX
+
 	// Restore user register state.
 	REGISTERS_LOAD(AX, 0)
 	MOVQ PTRACE_RIP(AX), CX    // Needed for SYSRET.
 	MOVQ PTRACE_FLAGS(AX), R11 // Needed for SYSRET.
-	MOVQ PTRACE_RSP(AX), SP    // Restore the stack directly.
-	MOVQ PTRACE_RAX(AX), AX    // Restore AX (scratch).
+
+	// restore userCR3, AX, SP
+	POPQ AX	                            // Get userCR3
+	BYTE $0x0f; BYTE $0x22; BYTE $0xd8; // Switch to userCR3 (MOVQ AX, CR3)
```

Comment:
Make this a macro? WRITE_CR3() ?

Reply:
Good suggestion. Before the final summit of the PR, I used difference registers in different place storing the userCR3. We might change it again if we go extreemely paranoid to use %rbp as the scratch register. Current code use %rax as the scratch register, so the %rax has to be stored on ENTRY_SCRATCH0, which is possible to be fetched by malicious APP. Leaking %rbp&%rsp is better than leaking %rax, if leaking %rax is a concern.

Reply:
Did it. Thank you.